internal/trie/node/hash.go

```diff
@@ -5,51 +5,78 @@ package node
 
 import (
 	"bytes"
+	"fmt"
 
 	"github.com/ChainSafe/gossamer/internal/trie/pools"
 	"github.com/ChainSafe/gossamer/lib/common"
 )
 
-// EncodeAndHash returns the encoding of the node and
+// MerkleValue produces the Merkle value from the encoding of a node.
+// For root nodes, the Merkle value is always the Blak2b hash of the encoding.
+// For other nodes, the Merkle value is either:
+// - the encoding if it is less than 32 bytes
+// - the Blake2b hash of the encoding
+func MerkleValue(encoding []byte, isRoot bool) (merkleValue []byte, err error) {
+	if !isRoot && len(encoding) < 32 {
+		merkleValue = make([]byte, len(encoding))
+		copy(merkleValue, encoding)
+		return merkleValue, nil
+	}
+
+	hashDigest, err := common.Blake2bHash(encoding)
+	if err != nil {
+		return nil, err
+	}
+
+	merkleValue = hashDigest[:]
+	return merkleValue, nil
+}
+
+// MerkleValue returns the encoding of the node and
 // the blake2b hash digest of the encoding of the node.
 // If the encoding is less than 32 bytes, the hash returned
 // is the encoding and not the hash of the encoding.
-func (n *Node) EncodeAndHash(isRoot bool) (encoding, hash []byte, err error) {
-	if !n.Dirty && n.Encoding != nil && n.HashDigest != nil {
-		return n.Encoding, n.HashDigest, nil
+func (n *Node) MerkleValue(isRoot bool) (merkleValue []byte, err error) {
+	if !n.Dirty && n.HashDigest != nil {
+		return n.HashDigest, nil
 	}
 
-	buffer := pools.EncodingBuffers.Get().(*bytes.Buffer)
-	buffer.Reset()
-	defer pools.EncodingBuffers.Put(buffer)
+	_, merkleValue, err = n.EncodeAndHash(isRoot)
+	return merkleValue, err
+}
 
-	err = n.Encode(buffer)
-	if err != nil {
-		return nil, nil, err
+// EncodeAndHash returns the encoding of the node and the
+// Merkle value of the node. See the `MerkleValue`
+// method for more details on the value of the Merkle value.
+func (n *Node) EncodeAndHash(isRoot bool) (encoding, merkleValue []byte, err error) {
+	if !n.Dirty && n.Encoding != nil && n.HashDigest != nil {
+		return n.Encoding, n.HashDigest, nil
 	}
 
-	bufferBytes := buffer.Bytes()
+	if n.Dirty || n.Encoding == nil {
+		buffer := pools.EncodingBuffers.Get().(*bytes.Buffer)
+		buffer.Reset()
+		defer pools.EncodingBuffers.Put(buffer)
 
-	// TODO remove this copying since it defeats the purpose of `buffer`
-	// and the sync.Pool.
-	n.Encoding = make([]byte, len(bufferBytes))
-	copy(n.Encoding, bufferBytes)
-	encoding = n.Encoding // no need to copy
+		err = n.Encode(buffer)
+		if err != nil {
+			return nil, nil, fmt.Errorf("encode node: %w", err)
+		}
 
-	if !isRoot && buffer.Len() < 32 {
-		n.HashDigest = make([]byte, len(bufferBytes))
-		copy(n.HashDigest, bufferBytes)
-		hash = n.HashDigest // no need to copy
-		return encoding, hash, nil
+		bufferBytes := buffer.Bytes()
+
+		// TODO remove this copying since it defeats the purpose of `buffer`
+		// and the sync.Pool.
```

Comment:
I'm curious about how copying defeats the purpose of `buffer` and the sync.Pool?

Reply:
That's just copy paste from older code.

Don't forget `bytes.Buffer` are really just byte slices.
If we copy from the buffer to a byte slice, the GC will keep track of the copied byte slice.
So we might just use a single byte slice (or a single local buffer) without the sync.pool buffer, to save the copying operation.

We should ideally address this later for the sake of performance, but it's not a priority.